### ISSUE TITLE
Writing in parallel to the same FBM

### ISSUE THREAD

**Issue by nbenn** [privefl/bigstatsr#67]
Do you know whether it is safe to write to separate regions of an `FBM` from multiple processes? In [big_copy()](https://github.com/privefl/bigstatsr/blob/5be45a729b707d155856a9a9a4801a61ea3dafe6/R/FBM-copy.R#L51) you have this comment, suggesting this is not a good idea. Are you sure about this? Or is it more about being cautious?

**Comment by privefl** [comment]
It is safe to write to SEPARATE regions at the same time.
But it doesn't usually accelerate things because writing to disk is not a parallel task.

**Comment by nbenn** [comment]
For me the reason to write in parallel is not speed, but not wanting to serialize data and send it to the master process.

I did run into issues though, MacOS kernel panicked twice (out of maybe 5 tries to write ~30Gb in parallel). Possibly it's due to the fact, that while my intention is to write to separate regions (i.e. non-overlapping matrix indices), the writes that actually do get carried out might overlap. If entire pages are written out (i.e. chunks of 4kb) this could happen as my writes are not aligned to pages.

Does this sound reasonable? Have you ever run into similar issues?

For the moment I solved this by serializing my writes using a mutex (from [synchronicity](https://cran.r-project.org/package=synchronicity).

**Comment by privefl** [comment]
I think paging should be taken care automatically by mio.

What do you mean panicked? Maybe it was just some `gc()` running?

I prefer using {flock} when I need locks.

**Comment by nbenn** [comment]
A kernel panic on macOS looks like

![panic](https://cdn.macpaw.com/uploads/images/kernel-panic-error.png)

and then the computer reboots. I got the following report

```
*** Panic Report ***
panic(cpu 2 caller 0xffffff800f7eba91): "Failed to restore dirty state on failure\n"@/BuildRoot/Library/Caches/com.apple.xbs/Sources/xnu/xnu-4903.251.3/bsd/vfs/vfs_cluster.c:6270
Backtrace (CPU 2), Frame : Return Address
0xffffff921f0fb970 : 0xffffff800f5af82d 
0xffffff921f0fb9c0 : 0xffffff800f6eb795 
0xffffff921f0fba00 : 0xffffff800f6dcf4a 
0xffffff921f0fba70 : 0xffffff800f55cae0 
0xffffff921f0fba90 : 0xffffff800f5af247 
0xffffff921f0fbbb0 : 0xffffff800f5af093 
0xffffff921f0fbc20 : 0xffffff800f7eba91 
0xffffff921f0fbc90 : 0xffffff800f7ec32c 
0xffffff921f0fbd40 : 0xffffff800f63dfca 
0xffffff921f0fbf10 : 0xffffff800f6dd54a 
0xffffff921f0fbfa0 : 0xffffff800f55c9ef 

BSD process name corresponding to current thread: R

Mac OS version:
18E226

Kernel version:
Darwin Kernel Version 18.5.0: Mon Mar 11 20:40:32 PDT 2019; root:xnu-4903.251.3~3/RELEASE_X86_64
...
```

How do you issue a lock from R? Are you using [flock](https://cran.r-project.org/package=flock)?

**Comment by privefl** [comment]
Hum, I never experienced this.
But, I don't use Mac.
If you manage to reproduce this problem with a simple example, please open another issue.

An example of using {flock}: https://privefl.github.io/blog/a-guide-to-parallelism-in-r/#advanced-parallelism-synchronization.

**Comment by nbenn** [comment]
Thanks for the `flock` reference. Ok, if I find an example that reproducibly triggers the above, I'll open a new issue. In the meantime, exclusive writes is probably the way to go. Thanks for the discussion.
